projects/example-app/src/app/auth/actions/auth-api.actions.ts

```diff
@@ -1,26 +1,17 @@
-import { Action } from '@ngrx/store';
+import { union, props, createAction } from '@ngrx/store';
 import { User } from '@example-app/auth/models/user';
 
-export enum AuthApiActionTypes {
-  LoginSuccess = '[Auth/API] Login Success',
-  LoginFailure = '[Auth/API] Login Failure',
-  LoginRedirect = '[Auth/API] Login Redirect',
-}
+export const loginSuccess = createAction(
+  '[Auth/API] Login Success',
+  props<{ user: User }>()
+);
 
-export class LoginSuccess implements Action {
-  readonly type = AuthApiActionTypes.LoginSuccess;
+export const loginFailure = createAction(
+  '[Auth/API] Login Failure',
+  props<{ error: any }>()
+);
 
-  constructor(public payload: { user: User }) {}
-}
+export const loginRedirect = createAction('[Auth/API] Login Redirect');
 
-export class LoginFailure implements Action {
-  readonly type = AuthApiActionTypes.LoginFailure;
-
-  constructor(public payload: { error: any }) {}
-}
-
-export class LoginRedirect implements Action {
-  readonly type = AuthApiActionTypes.LoginRedirect;
-}
-
-export type AuthApiActionsUnion = LoginSuccess | LoginFailure | LoginRedirect;
+const all = union({ loginSuccess, loginFailure, loginRedirect });
```

Comment:
Why not combine these two lines into one?

Reply:
Unfortunately, because of https://github.com/Microsoft/TypeScript/issues/6606 they cannot be combined (at least I didn't find an easy way to do it).

Alternatively, instead of 
```ts
 const all = union({ loginSuccess, loginFailure, loginRedirect });
export type AuthApiActionsUnion = typeof all;
```
we can do the following
```
// drop the union(...) altogether
export type AuthApiActionsUnion = ReturnType<
  typeof loginSuccess | typeof loginFailure | typeof loginRedirect
>;
```
This is basically what `union` fn does anyway.
 
WDYT?

Reply:
The second alternative would be to export the `union` itself and use `typeof AuthApiActionsUnion` instead of `AuthApiActionsUnion`.
That's the best alternative from my point of view, but it might introduce some friction

Reply:
I'd prefer less friction with the union approach you already had

Reply:
Hmm, what about the following?

```ts
export const CollectionApiActions = {
  addBookSuccess: createAction(
    '[Collection/API] Add Book Success',
    props<{ book: Book }>()
  ),
  addBookFailure: createAction('[Collection/API] Add Book Failure'),
  loadBooksFailure,
};

const all = union(CollectionApiActions);
export type CollectionApiActionsUnion = typeof all;
```

At the barrel export in `index.ts` it would be:
```diff
-- import * as CollectionApiActions from './collection-api.actions';
++ export * from './collection-api.actions';
```

Reply:
Personally, I don't see the need to wrap action creators with an object - that does feel like it saves us much :)
I typically do module import anyway and we don't use the barrel (aka entry-point) modules - they are OK for libraries, but for regular project they just add overhead without clear benefit, IMHO.

Reply:
What I ended up doing is use the `union` function when there are 2+ action creators (except a single case, where I added the comments) within a file and just do the `ReturnType<typeof foo>` when there's only one.